core/server/proxy/src/main/java/alluxio/proxy/s3/signature/AwsSignatureProcessor.java

```diff
@@ -0,0 +1,229 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.proxy.s3.signature;
+
+import static alluxio.proxy.s3.signature.SignerConstants.X_AMZ_DATE;
+
+import alluxio.proxy.s3.S3ErrorCode;
+import alluxio.proxy.s3.S3Exception;
+import alluxio.proxy.s3.auth.AwsAuthInfo;
+import alluxio.proxy.s3.signature.utils.AwsAuthV2HeaderParserUtils;
+import alluxio.proxy.s3.signature.utils.AwsAuthV4HeaderParserUtils;
+import alluxio.proxy.s3.signature.utils.AwsAuthV4QueryParserUtils;
+
+import org.apache.commons.lang3.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.ws.rs.container.ContainerRequestContext;
+import javax.ws.rs.core.MultivaluedMap;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Parser to process AWS V2 & V4 auth request. Creates string to sign and auth
+ * header. For more details refer to AWS documentation https://docs.aws
+ * .amazon.com/general/latest/gr/sigv4-create-canonical-request.html.
+ **/
+
+public class AwsSignatureProcessor {
+  private static final Logger LOG =
+            LoggerFactory.getLogger(AwsSignatureProcessor.class);
+  private static final String AUTHORIZATION = "Authorization";
+
+  private ContainerRequestContext mContext;
+
+  /**
+   * Create a new {@link AwsSignatureProcessor}.
+   *
+   * @param context ContainerRequestContext
+   */
+  public AwsSignatureProcessor(ContainerRequestContext context) {
+    mContext = context;
+  }
+
+  /**
+   * Extract signature info from request.
+   * @return SignatureInfo
+   * @throws S3Exception
+   */
+  public SignatureInfo parseSignature() throws S3Exception {
+    LowerCaseKeyStringMap headers =
+        LowerCaseKeyStringMap.fromHeaderMap(mContext.getHeaders());
+
+    String authHeader = headers.get(AUTHORIZATION);
+    String dateHeader = headers.get(X_AMZ_DATE);
+    Map<String, String> queryParameters = StringToSignProducer.fromMultiValueToSingleValueMap(
+            mContext.getUriInfo().getQueryParameters());
+
+    SignatureInfo signatureInfo = null;
+    if ((signatureInfo =
+            AwsAuthV2HeaderParserUtils.parseSignature(authHeader)) != null
+        || (signatureInfo =
+            AwsAuthV4HeaderParserUtils.parseSignature(authHeader, dateHeader)) != null
+        || (signatureInfo =
+            AwsAuthV4QueryParserUtils.parseSignature(queryParameters)) != null) {
+      return signatureInfo;
+    } else {
+      return new SignatureInfo(
```

Comment:
Might worth having some simpler constructor in `SignatureInfo` that just takes fewer parameters. Up to you.

Reply:
finish.